In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path

In [6]:
org_FSStats_data = pd.read_excel("../data/org_FSStats.xlsx")

print("Rows and columns:", org_FSStats_data.shape)
print("\nColumn names:")
print(org_FSStats_data.columns.tolist())

org_FSStats_data.head()

Rows and columns: (28287, 58)

Column names:
['__ID', '_id_site', '_Year', '_Month', 'DateForSearch', 'zzCreationTS', 'zzModificationTS', 'AdultsReg', 'ChildrenReg', 'HouseholdsReg', 'IndividualsReg', 'Meals', 'Misc', 'People', 'PoundsReg', 'SeniorsReg', 'Snacks', 'UniqueVisitsReg', 'County', 'FederalDistrict', 'StateDistrict', 'FoodBank', 'ChildrenExtra', 'ChildrenProduce', 'ChildrenHoliday', 'AdultsExtra', 'AdultsProduce', 'AdultsHoliday', 'IndividualsExtra', 'SeniorsExtra', 'SeniorsProduce', 'SeniorsHoliday', 'IndividualsProduce', 'IndividualsHoliday', 'UniqueVisitsExtra', 'UniqueVisitsProduce', 'UniqueVisitsHoliday', 'UniqueVisitsTotal', 'SeniorsTotal', 'IndividualsTotal', 'ChildrenTotal', 'AdultsTotal', 'HouseholdsExtra', 'HouseholdsProduce', 'HouseholdsHoliday', 'HouseholdsTotal', 'PoundsExtra', 'PoundsProduce', 'PoundsHoliday', 'PoundsTotal', 'AgencyType', '_id_bank', 'SuperShelf', '_YearMonth', 'PoundsPerIndiv', 'IndividualsPerHH', 'id_agency', 'PoundsPerHH']


,__ID,_id_site,_Year,_Month,DateForSearch,zzCreationTS,zzModificationTS,AdultsReg,ChildrenReg,HouseholdsReg,...,PoundsHoliday,PoundsTotal,AgencyType,_id_bank,SuperShelf,_YearMonth,PoundsPerIndiv,IndividualsPerHH,id_agency,PoundsPerHH
0,2753,1254.0,2022,1,2022-01-01,2022-01-10T21:48:14Z,2025-12-30T13:50:18Z,1519.0,1048.0,1027.0,...,NaN,41548.0,Brick and Mortar Site,6.0,NaN,202201,15.0,3.0,NaN,40.0
1,2762,14434.0,2022,1,2022-01-01,2022-01-10T15:01:29Z,2025-12-30T13:50:18Z,23.0,16.0,17.0,...,NaN,900.0,Brick and Mortar Site,1.0,NaN,202201,19.0,3.0,14434.0,53.0
2,12862,1069.0,2022,3,2022-03-01,2022-03-02T16:18:49Z,2026-09-02T20:20:24Z,281.0,27.0,220.0,...,NaN,21345.0,Food Shelf,14179.0,NaN,202203,62.0,2.0,16988.0,97.0
3,12863,1069.0,2022,1,2022-01-01,2022-03-08T09:50:31Z,2026-09-02T20:20:24Z,281.0,21.0,216.0,...,NaN,15684.0,Food Shelf,14179.0,NaN,202201,48.0,2.0,16988.0,73.0
4,12864,14680.1,2022,1,2022-01-01,2022-03-08T09:50:31Z,2025-12-30T13:50:24Z,142.0,93.0,91.0,...,NaN,5761.0,Tribal Program,14179.0,NaN,202201,22.0,3.0,NaN,63.0


In [7]:
print("Shape:", org_FSStats_data.shape)

print("\nColumns:")
print(org_FSStats_data.columns.tolist())

display(org_FSStats_data.head())

print("\nMissing values:")
print(org_FSStats_data.isnull().sum())

Shape: (28287, 58)

Columns:
['__ID', '_id_site', '_Year', '_Month', 'DateForSearch', 'zzCreationTS', 'zzModificationTS', 'AdultsReg', 'ChildrenReg', 'HouseholdsReg', 'IndividualsReg', 'Meals', 'Misc', 'People', 'PoundsReg', 'SeniorsReg', 'Snacks', 'UniqueVisitsReg', 'County', 'FederalDistrict', 'StateDistrict', 'FoodBank', 'ChildrenExtra', 'ChildrenProduce', 'ChildrenHoliday', 'AdultsExtra', 'AdultsProduce', 'AdultsHoliday', 'IndividualsExtra', 'SeniorsExtra', 'SeniorsProduce', 'SeniorsHoliday', 'IndividualsProduce', 'IndividualsHoliday', 'UniqueVisitsExtra', 'UniqueVisitsProduce', 'UniqueVisitsHoliday', 'UniqueVisitsTotal', 'SeniorsTotal', 'IndividualsTotal', 'ChildrenTotal', 'AdultsTotal', 'HouseholdsExtra', 'HouseholdsProduce', 'HouseholdsHoliday', 'HouseholdsTotal', 'PoundsExtra', 'PoundsProduce', 'PoundsHoliday', 'PoundsTotal', 'AgencyType', '_id_bank', 'SuperShelf', '_YearMonth', 'PoundsPerIndiv', 'IndividualsPerHH', 'id_agency', 'PoundsPerHH']


,__ID,_id_site,_Year,_Month,DateForSearch,zzCreationTS,zzModificationTS,AdultsReg,ChildrenReg,HouseholdsReg,...,PoundsHoliday,PoundsTotal,AgencyType,_id_bank,SuperShelf,_YearMonth,PoundsPerIndiv,IndividualsPerHH,id_agency,PoundsPerHH
0,2753,1254.0,2022,1,2022-01-01,2022-01-10T21:48:14Z,2025-12-30T13:50:18Z,1519.0,1048.0,1027.0,...,NaN,41548.0,Brick and Mortar Site,6.0,NaN,202201,15.0,3.0,NaN,40.0
1,2762,14434.0,2022,1,2022-01-01,2022-01-10T15:01:29Z,2025-12-30T13:50:18Z,23.0,16.0,17.0,...,NaN,900.0,Brick and Mortar Site,1.0,NaN,202201,19.0,3.0,14434.0,53.0
2,12862,1069.0,2022,3,2022-03-01,2022-03-02T16:18:49Z,2026-09-02T20:20:24Z,281.0,27.0,220.0,...,NaN,21345.0,Food Shelf,14179.0,NaN,202203,62.0,2.0,16988.0,97.0
3,12863,1069.0,2022,1,2022-01-01,2022-03-08T09:50:31Z,2026-09-02T20:20:24Z,281.0,21.0,216.0,...,NaN,15684.0,Food Shelf,14179.0,NaN,202201,48.0,2.0,16988.0,73.0
4,12864,14680.1,2022,1,2022-01-01,2022-03-08T09:50:31Z,2025-12-30T13:50:24Z,142.0,93.0,91.0,...,NaN,5761.0,Tribal Program,14179.0,NaN,202201,22.0,3.0,NaN,63.0



Missing values:
__ID                       0
_id_site                   0
_Year                      0
_Month                     0
DateForSearch              0
zzCreationTS               0
zzModificationTS           0
AdultsReg               2062
ChildrenReg             2186
HouseholdsReg           2238
IndividualsReg          2027
Meals                  15481
Misc                   28287
People                 28287
PoundsReg               2235
SeniorsReg              2236
Snacks                 28287
UniqueVisitsReg        28268
County                   132
FederalDistrict          359
StateDistrict            516
FoodBank                  36
ChildrenExtra          28287
ChildrenProduce        28287
ChildrenHoliday        28287
AdultsExtra            28287
AdultsProduce          28287
AdultsHoliday          28287
IndividualsExtra       28287
SeniorsExtra           28287
SeniorsProduce         28287
SeniorsHoliday         28287
IndividualsProduce     28287
IndividualsHoliday     282

In [8]:
# Select the important variables for Q6
q6_columns = [
    "_id_site",
    "_Year",
    "_Month",
    "County",
    "AgencyType",
    "HouseholdsReg",
    "IndividualsReg",
    "PoundsReg"
]

q6_data = org_FSStats_data[q6_columns].copy()

display(q6_data.head(10))

,_id_site,_Year,_Month,County,AgencyType,HouseholdsReg,IndividualsReg,PoundsReg
0,1254.0,2022,1,St. Louis,Brick and Mortar Site,1027.0,2706.0,41548.0
1,14434.0,2022,1,Olmsted,Brick and Mortar Site,17.0,47.0,900.0
2,1069.0,2022,3,Hennepin-Minneapolis,Food Shelf,220.0,343.0,21345.0
3,1069.0,2022,1,Hennepin-Minneapolis,Food Shelf,216.0,330.0,15684.0
4,14680.1,2022,1,Koochiching,Tribal Program,91.0,266.0,5761.0
5,1097.0,2022,1,Hennepin-Minneapolis,Food Shelf,848.0,1677.0,49570.0
6,1091.0,2022,1,Hennepin-Minneapolis,Food Shelf,337.0,1632.0,26520.0
7,14711.1,2022,1,Renville,Food Shelf,410.0,1287.0,14368.0
8,1083.0,2022,1,Hennepin-Suburbs,Food Shelf,326.0,674.0,22434.0
9,1086.0,2022,1,Hennepin-Suburbs,Food Shelf,822.0,3134.0,994394.0


In [11]:
print("Year:")
print(q6_data["_Year"].value_counts().sort_index())
print("Months:")
print(q6_data["_Month"].value_counts().sort_index())
print("AgencyType:")
print(q6_data["AgencyType"].value_counts().sort_index())

Year:
_Year
2022    5959
2023    6569
2024    6228
2025    6023
2026    3508
Name: count, dtype: int64
Months:
_Month
0       25
1     2522
2     2553
3     2552
4     2516
5     2542
6     2533
7     2575
8     2062
9     2096
10    2105
11    2103
12    2102
15       1
Name: count, dtype: int64
AgencyType:
AgencyType
Agency or Tribal Nation       17
Brick and Mortar Site      14751
Brick and Mortar Sites         9
College Campus Program       165
Food Shelf                  7406
Meal Program                1593
Meal Programs                 89
Mobile                       354
On-Site Meal Programs       1040
Other Non-Profit              22
Outreach Program            2480
Tribal Program               223
Name: count, dtype: int64


In [13]:
# Find records with invalid months
invalid_months = org_FSStats_data[
    ~org_FSStats_data["_Month"].between(1, 12)
]

display(invalid_months[
    [
        "_id_site",
        "_Year",
        "_Month",
        "_YearMonth",
        "DateForSearch",
        "County",
        "AgencyType"
    ]
])

print("Invalid records:", len(invalid_months))

agency_year = pd.crosstab(
    q6_data["_Year"],
    q6_data["AgencyType"]
)

display(agency_year)

,_id_site,_Year,_Month,_YearMonth,DateForSearch,County,AgencyType
10251,1148.0,2023,15,202315,2024-03-01,Polk,Brick and Mortar Site
12369,14810.0,2023,0,202300,2022-12-01,St. Louis,Tribal Program
12370,1072.0,2023,0,202300,2022-12-01,St. Louis,Brick and Mortar Site
12371,14717.0,2023,0,202300,2022-12-01,Carlton,Outreach Program
12372,1103.0,2023,0,202300,2022-12-01,St. Louis,Brick and Mortar Site
12373,1088.0,2023,0,202300,2022-12-01,St. Louis,Brick and Mortar Site
12374,1110.0,2023,0,202300,2022-12-01,Carlton,Brick and Mortar Site
12375,1108.0,2023,0,202300,2022-12-01,St. Louis,Brick and Mortar Site
12376,1250.0,2023,0,202300,2022-12-01,St. Louis,Brick and Mortar Site
12377,1410.0,2023,0,202300,2022-12-01,St. Louis,Brick and Mortar Site


Invalid records: 26


AgencyType,Agency or Tribal Nation,Brick and Mortar Site,Brick and Mortar Sites,College Campus Program,Food Shelf,Meal Program,Meal Programs,Mobile,On-Site Meal Programs,Other Non-Profit,Outreach Program,Tribal Program
_Year,,,,,,,,,,,,
2022,0,1355,0,0,3649,132,0,120,481,11,160,47
2023,0,1416,1,0,3757,144,1,234,559,11,349,49
2024,0,4721,8,60,0,454,88,0,0,0,771,48
2025,3,4586,0,63,0,554,0,0,0,0,766,48
2026,14,2673,0,42,0,309,0,0,0,0,434,31


In [14]:
# Separate valid and invalid months
valid = org_FSStats_data[
    org_FSStats_data["_Month"].between(1, 12)
].copy()

invalid = org_FSStats_data[
    ~org_FSStats_data["_Month"].between(1, 12)
].copy()

# Convert the date column
valid["DateForSearch"] = pd.to_datetime(valid["DateForSearch"])
invalid["DateForSearch"] = pd.to_datetime(invalid["DateForSearch"])

# Check whether an invalid record has a matching
# site and date among valid records
check = invalid.merge(
    valid[["_id_site", "DateForSearch"]],
    on=["_id_site", "DateForSearch"],
    how="left",
    indicator=True
)

print(check["_merge"].value_counts())

display(
    check[[
        "_id_site",
        "_Year",
        "_Month",
        "DateForSearch",
        "_merge"
    ]]
)

_merge
both          26
left_only      0
right_only     0
Name: count, dtype: int64


,_id_site,_Year,_Month,DateForSearch,_merge
0,1148.0,2023,15,2024-03-01,both
1,14810.0,2023,0,2022-12-01,both
2,1072.0,2023,0,2022-12-01,both
3,14717.0,2023,0,2022-12-01,both
4,1103.0,2023,0,2022-12-01,both
5,1088.0,2023,0,2022-12-01,both
6,1110.0,2023,0,2022-12-01,both
7,1108.0,2023,0,2022-12-01,both
8,1250.0,2023,0,2022-12-01,both
9,1410.0,2023,0,2022-12-01,both


In [15]:
# Compare the three important Q6 variables
activity_cols = [
    "HouseholdsReg",
    "IndividualsReg",
    "PoundsReg"
]

# Match suspicious records with valid records
comparison = invalid.merge(
    valid[
        ["_id_site", "DateForSearch"] + activity_cols
    ],
    on=["_id_site", "DateForSearch"],
    how="left",
    suffixes=("_invalid", "_valid")
)

# Compare activity values
for col in activity_cols:
    comparison[f"{col}_same"] = (
        comparison[f"{col}_invalid"].fillna(-1)
        ==
        comparison[f"{col}_valid"].fillna(-1)
    )

# Check if all three variables match
comparison["all_match"] = comparison[
    [f"{col}_same" for col in activity_cols]
].all(axis=1)

print("Total comparisons:", len(comparison))
print("\nActivity values match:")
print(comparison["all_match"].value_counts())

display(comparison[
    ["_id_site", "DateForSearch", "all_match"]
])

Total comparisons: 26

Activity values match:
all_match
False    26
Name: count, dtype: int64


,_id_site,DateForSearch,all_match
0,1148.0,2024-03-01,False
1,14810.0,2022-12-01,False
2,1072.0,2022-12-01,False
3,14717.0,2022-12-01,False
4,1103.0,2022-12-01,False
5,1088.0,2022-12-01,False
6,1110.0,2022-12-01,False
7,1108.0,2022-12-01,False
8,1250.0,2022-12-01,False
9,1410.0,2022-12-01,False


In [16]:
# Check which activity variables differ
for col in activity_cols:
    print(f"\n{col}:")
    print(comparison[f"{col}_same"].value_counts())

# Display original and matching values side by side
columns = ["_id_site", "DateForSearch"]

for col in activity_cols:
    columns.extend([
        f"{col}_invalid",
        f"{col}_valid"
    ])

display(comparison[columns])


HouseholdsReg:
HouseholdsReg_same
False    25
True      1
Name: count, dtype: int64

IndividualsReg:
IndividualsReg_same
False    26
Name: count, dtype: int64

PoundsReg:
PoundsReg_same
False    26
Name: count, dtype: int64


,_id_site,DateForSearch,HouseholdsReg_invalid,HouseholdsReg_valid,IndividualsReg_invalid,IndividualsReg_valid,PoundsReg_invalid,PoundsReg_valid
0,1148.0,2024-03-01,33.0,43.0,69.0,97.0,2043.0,2902.4
1,14810.0,2022-12-01,54.0,1129.0,129.0,1129.0,1785.0,9285.0
2,1072.0,2022-12-01,183.0,165.0,421.0,373.0,8778.0,10796.0
3,14717.0,2022-12-01,332.0,0.0,755.0,0.0,11625.0,0.0
4,1103.0,2022-12-01,274.0,362.0,691.0,962.0,12128.0,12033.0
5,1088.0,2022-12-01,691.0,565.0,1106.0,974.0,40515.0,36595.0
6,1110.0,2022-12-01,146.0,106.0,463.0,310.0,13535.0,8151.0
7,1108.0,2022-12-01,97.0,82.0,202.0,186.0,5820.0,5740.0
8,1250.0,2022-12-01,118.0,71.0,188.0,151.0,6948.0,NaN
9,1410.0,2022-12-01,653.0,409.0,1772.0,1053.0,37695.0,27830.0


In [ ]:
# Keep the invalid records separately
invalid_records = org_FSStats_data[
    ~org_FSStats_data["_Month"].between(1, 12)
].copy()

# Select valid 2024 records
q6_2024 = org_FSStats_data[
    (org_FSStats_data["_Year"] == 2024) &
    (org_FSStats_data["_Month"].between(1, 12))
].copy()

print("2024 observations:", len(q6_2024))

print("Unique sites:", q6_2024["_id_site"].nunique())

print("Unique counties:", q6_2024["County"].nunique())

print("\nAgency types:")
print(q6_2024["AgencyType"].value_counts())

###Minnesota has 87counties !!!! data shows 90

2024 observations: 6228
Unique sites: 575
Unique counties: 90

Agency types:
AgencyType
Brick and Mortar Site     4721
Outreach Program           771
Meal Program               454
Meal Programs               88
College Campus Program      60
Tribal Program              48
Brick and Mortar Sites       8
Name: count, dtype: int64


In [18]:
# Display all county names and their frequencies
county_counts = q6_2024["County"].value_counts(
    dropna=False
).sort_index()

print(county_counts.to_string())

# Check missing county information
print("\nMissing county records:")
print(q6_2024["County"].isna().sum())

County
Aitkin                   59
Anoka                   144
Becker                   48
Beltrami                143
Benton                   36
Big Stone                12
Blue Earth               53
Brown                    36
Carlton                  48
Carver                   59
Cass                    106
Chippewa                 12
Chisago                  60
Clay                    193
Clearwater               12
Cook                     36
Cottonwood               36
Crow Wing                94
Dakota                  250
Dodge                    24
Douglas                  12
Faribault                31
Fillmore                 56
Freeborn                 48
Goodhue                  57
Grant                    24
Hennepin-Minneapolis    925
Hennepin-Suburbs        352
Houston                  44
Hubbard                  48
Isanti                   24
Itasca                  115
Jackson                  36
Kanabec                  36
Kandiyohi                36
Kittson      

In [20]:
# Create a copy so we preserve the original dataset
q6_clean = org_FSStats_data.copy()

# Standardize county names
county_mapping = {
    "Hennepin-Minneapolis": "Hennepin",
    "Hennepin-Suburbs": "Hennepin",
    "Ramsey-Saint Paul": "Ramsey",
    "Ramsey-Suburbs": "Ramsey",
    "Saint Louis": "St. Louis"
}

q6_clean["County"] = (
    q6_clean["County"]
    .str.strip()
    .replace(county_mapping)
)

# Examine the 2024 county names
q6_2024_clean = q6_clean[q6_clean["_Year"] == 2024]

print("Unique counties:",
      q6_2024_clean["County"].nunique())

print("Missing counties:",
      q6_2024_clean["County"].isna().sum())

#investigating the misssing county data
# Select records with missing county information
missing_county = q6_2024_clean[
    q6_2024_clean["County"].isna()
]

# Check the types of organizations involved
print("Agency types with missing county:")
print(missing_county["AgencyType"].value_counts(dropna=False))

# Examine site IDs and activity
display(missing_county[
    [
        "_id_site",
        "AgencyType",
        "FoodBank",
        "HouseholdsReg",
        "IndividualsReg",
        "PoundsReg"
    ]
].head(15))

Unique counties: 86
Missing counties: 64
Agency types with missing county:
AgencyType
NaN    64
Name: count, dtype: int64


,_id_site,AgencyType,FoodBank,HouseholdsReg,IndividualsReg,PoundsReg
12711,15364.0,NaN,Second Harvest Heartland,148.0,393.0,5556.0
12714,15366.0,NaN,Second Harvest Heartland,4188.0,12356.0,180317.0
12715,15361.0,NaN,Second Harvest Heartland,620.0,1700.0,35948.0
12779,15365.0,NaN,Second Harvest Heartland,913.0,2353.0,58330.0
12956,15389.0,NaN,NaN,NaN,NaN,NaN
13063,15364.0,NaN,Second Harvest Heartland,256.0,996.0,8184.0
13109,15366.0,NaN,Second Harvest Heartland,3961.0,11423.0,168823.0
13110,15365.0,NaN,Second Harvest Heartland,856.0,2234.0,56716.0
13111,15361.0,NaN,Second Harvest Heartland,570.0,1451.0,33360.0
13403,15426.0,NaN,NaN,44.0,133.0,3728.0


In [21]:
org_data = pd.read_excel("../data/Organization.xlsx")

print("Shape:", org_data.shape)
print("\nColumns:")
print(org_data.columns.tolist())

org_data.head()

Shape: (1202, 68)

Columns:
['__ID', 'CreateTime', 'zzCreationTS', 'zzModificationTS', '_id_type', 'City', 'State', 'Zip', 'MailCity', 'MailState', 'MailZip', 'Eligibility', 'Conditions', 'MarketSeasons', 'zcountDaysOperations', 'Region', 'SuperRegion', 'County', 'StateDistrict', 'FederalDistrict', 'DistrictSection', 'AgencyType', '_id_subtype', 'AgencySubType', 'isFoodBank', 'isAgency', 'isDistributionSite', 'HoldsInventoryForOthers', 'isWebscmShipTo', 'WebscmShipToName', 'StateClass', 'MFS', 'MCS', 'cCountSites', 'isActive', 'ActiveInactive', 'InactiveDate', 'TaxEINVerifiedYN', 'TaxEINVerifiedDate', 'TaxEINVerifiedMethod', 'isTEFAPEligible', 'isTEFAPEligible_Alert', 'TefapStatus', 'TefapActiveDate', 'TefapInActiveDate', 'TEFAPElligibilityMeans', 'SuperShelfStatus', 'SuperShelfCertDate', '_id_Bank', '_id_Agency', '_id_IfFoodBank', 'isFindHelpMap', 'FindHelpMapCategory', '_id_findhelpMapCat', 'z12MoAverage_children', 'z12MoAverage_Adults', 'z12MoAverage_Senior', 'z12MoAverage_Indiv', '

,__ID,CreateTime,zzCreationTS,zzModificationTS,_id_type,City,State,Zip,MailCity,MailState,...,z12MoAverage_House,z12MoAverage_Pounds,z12MoAverage_LlbsPerPerson,z12MoAverage_PeoplePerHouse,z12MoAverage_TEFAPCases,z12MoAverage_LlbsPerHH,z12MoAverage_Meals,zNameIfActiveBank,block_geoid,is_zip_centroid
0,1.0,System.Byte[],NaN,2026-05-15T19:15:18Z,11.0,Rochester,MN,55904.0,Rochester,MN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Channel One Regional Food Bank*,2.710900e+14,0.0
1,3.0,System.Byte[],NaN,2026-08-06T10:27:10Z,1.0,Brooklyn Park,MN,55428.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Second Harvest Heartland,2.705303e+14,0.0
2,5.0,System.Byte[],NaN,2026-09-01T02:10:43Z,1.0,East Grand Forks,MN,56721.0,East Grand Forks,MN,...,NaN,NaN,NaN,NaN,530.0,NaN,NaN,North Country Food Bank Inc.,2.711902e+14,0.0
3,6.0,System.Byte[],NaN,2026-09-03T10:32:32Z,1.0,Duluth,MN,55808.0,Duluth,MN,...,NaN,NaN,NaN,NaN,108.0,NaN,NaN,Second Harvest Northland,2.713700e+14,0.0
4,7.0,System.Byte[],NaN,2026-01-07T13:35:05Z,1.0,Fargo,ND,58102.0,Fargo,ND,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.801700e+14,0.0


In [ ]:
#using or_data  to fix missing values in orgFSS_data 
#checking for similar features in Org_data
[col for col in org_data.columns
 if any(word in col.lower() for word in
        ["site", "county", "agency", "type", "geoid"])]

['_id_type',
 'County',
 'AgencyType',
 '_id_subtype',
 'AgencySubType',
 'isAgency',
 'isDistributionSite',
 'cCountSites',
 '_id_Agency',
 'block_geoid']

In [24]:
# Check how many activity site IDs appear in Organization.xlsx
site_ids_activity = set(q6_clean["_id_site"].dropna())
site_ids_org = set(org_data["__ID"].dropna())

matched_ids = site_ids_activity.intersection(site_ids_org)

print("Unique site IDs in activity:", len(site_ids_activity))
print("Unique IDs in organization:", len(site_ids_org))
print("Matching IDs:", len(matched_ids))
print("Match rate:",
      len(matched_ids) / len(site_ids_activity))


missing_site_ids = missing_county["_id_site"].dropna().unique()

org_matches = org_data[
    org_data["__ID"].isin(missing_site_ids)
][
    [
        "__ID",
        "County",
        "AgencyType",
        "AgencySubType",
        "isDistributionSite",
        "isActive",
        "block_geoid"
    ]
]

display(org_matches)
print("Sites recovered:", org_matches["__ID"].nunique())
print("Missing sites we started with:", len(missing_site_ids))

Unique site IDs in activity: 665
Unique IDs in organization: 1202
Matching IDs: 631
Match rate: 0.9488721804511279


,__ID,County,AgencyType,AgencySubType,isDistributionSite,isActive,block_geoid
737,15361.0,NaN,NaN,NaN,1.0,NaN,NaN
740,15364.0,NaN,NaN,NaN,1.0,NaN,NaN
741,15365.0,NaN,NaN,NaN,1.0,NaN,NaN
742,15366.0,NaN,NaN,NaN,1.0,NaN,NaN


Sites recovered: 4
Missing sites we started with: 38


In [25]:
# Site IDs that have missing county in 2024
missing_site_ids = missing_county["_id_site"].dropna().unique()

# Look for those same sites anywhere in the full activity dataset
history_lookup = org_FSStats_data[
    org_FSStats_data["_id_site"].isin(missing_site_ids)
][
    [
        "_id_site",
        "_Year",
        "County",
        "AgencyType",
        "FoodBank"
    ]
].sort_values(["_id_site", "_Year"])

display(history_lookup)

site_recovery = (
    history_lookup
    .groupby("_id_site")
    .agg(
        county_values=("County", lambda x: list(x.dropna().unique())),
        agency_values=("AgencyType", lambda x: list(x.dropna().unique()))
    )
    .reset_index()
)

display(site_recovery)

print("Sites with recoverable county:",
      site_recovery["county_values"].apply(len).gt(0).sum())

print("Sites with recoverable agency type:",
      site_recovery["agency_values"].apply(len).gt(0).sum())

,_id_site,_Year,County,AgencyType,FoodBank
9497,15361.0,2023,NaN,NaN,Second Harvest Heartland
9923,15361.0,2023,NaN,NaN,Second Harvest Heartland
10770,15361.0,2023,NaN,NaN,Second Harvest Heartland
11259,15361.0,2023,NaN,NaN,Second Harvest Heartland
11795,15361.0,2023,NaN,NaN,Second Harvest Heartland
...,...,...,...,...,...
13425,15430.0,2024,NaN,NaN,NaN
13416,15431.0,2024,NaN,NaN,NaN
13431,15432.0,2024,NaN,NaN,NaN
13406,15433.0,2024,NaN,NaN,NaN


,_id_site,county_values,agency_values
0,15361.0,[],[]
1,15364.0,[],[]
2,15365.0,[],[]
3,15366.0,[],[]
4,15383.0,[],[]
5,15388.0,[],[]
6,15389.0,[],[]
7,15390.0,[],[]
8,15393.0,[],[]
9,15398.0,[],[]


Sites with recoverable county: 0
Sites with recoverable agency type: 0


In [26]:
# Look at the agency IDs for the 38 problem sites
missing_agency = missing_county[
    ["_id_site", "id_agency", "_id_bank", "FoodBank"]
].drop_duplicates()

display(missing_agency)

print("Missing id_agency:",
      missing_agency["id_agency"].isna().sum())

print("Unique agency IDs:",
      missing_agency["id_agency"].nunique())

agency_matches = missing_agency.merge(
    org_data[
        [
            "_id_Agency",
            "County",
            "AgencyType",
            "AgencySubType",
            "City",
            "Zip",
            "isDistributionSite",
            "block_geoid"
        ]
    ],
    left_on="id_agency",
    right_on="_id_Agency",
    how="left"
)

print("Rows with county recovered:",
      agency_matches["County"].notna().sum())

print("Unique problem sites with county recovered:",
      agency_matches.loc[
          agency_matches["County"].notna(),
          "_id_site"
      ].nunique())

display(
    agency_matches[
        [
            "_id_site",
            "id_agency",
            "County",
            "AgencyType",
            "City",
            "Zip",
            "block_geoid"
        ]
    ]
)

,_id_site,id_agency,_id_bank,FoodBank
12711,15364.0,15364.0,3.0,Second Harvest Heartland
12714,15366.0,15366.0,3.0,Second Harvest Heartland
12715,15361.0,15361.0,3.0,Second Harvest Heartland
12779,15365.0,15365.0,3.0,Second Harvest Heartland
12956,15389.0,NaN,NaN,NaN
13403,15426.0,NaN,NaN,NaN
13404,15418.0,NaN,NaN,NaN
13405,15428.0,NaN,NaN,NaN
13406,15433.0,NaN,NaN,NaN
13407,15424.0,NaN,NaN,NaN


Missing id_agency: 34
Unique agency IDs: 4
Rows with county recovered: 17578
Unique problem sites with county recovered: 34


,_id_site,id_agency,County,AgencyType,City,Zip,block_geoid
0,15364.0,15364.0,NaN,NaN,NaN,NaN,NaN
1,15366.0,15366.0,NaN,NaN,NaN,NaN,NaN
2,15361.0,15361.0,NaN,NaN,NaN,NaN,NaN
3,15365.0,15365.0,NaN,NaN,NaN,NaN,NaN
4,15389.0,NaN,Saint Louis,Food Bank,Duluth,55808.0,2.713700e+14
...,...,...,...,...,...,...,...
19243,15411.0,NaN,Dakota,Other,NaN,55337.0,2.703706e+14
19244,15411.0,NaN,Chisago,NaN,NaN,55045.0,2.702511e+14
19245,15411.0,NaN,NaN,NaN,NaN,NaN,NaN
19246,15411.0,NaN,Hennepin-Minneapolis,Agency or Tribal Nation,Minneapolis,55409.0,2.705311e+14


In [27]:
# Only keep problem rows with a REAL agency ID
missing_agency_valid = missing_agency[
    missing_agency["id_agency"].notna()
].copy()

# Only keep Organization rows with a REAL agency ID
org_agency_valid = org_data[
    org_data["_id_Agency"].notna()
].copy()

agency_matches = missing_agency_valid.merge(
    org_agency_valid[
        [
            "_id_Agency",
            "County",
            "AgencyType",
            "AgencySubType",
            "City",
            "Zip",
            "isDistributionSite",
            "block_geoid"
        ]
    ],
    left_on="id_agency",
    right_on="_id_Agency",
    how="left"
)

print("Rows after correct merge:", len(agency_matches))

print("Problem sites with county recovered:",
      agency_matches.loc[
          agency_matches["County"].notna(),
          "_id_site"
      ].nunique())

display(agency_matches)

# Activity from unresolved-county rows
unresolved = q6_2024_clean[
    q6_2024_clean["County"].isna()
]

metrics = ["HouseholdsReg", "IndividualsReg", "PoundsReg"]

for col in metrics:
    unresolved_sum = unresolved[col].sum()
    total_sum = q6_2024_clean[col].sum()

    print(
        f"{col}: "
        f"{unresolved_sum:,.0f} unresolved out of {total_sum:,.0f} total "
        f"({100 * unresolved_sum / total_sum:.2f}%)"
    )

Rows after correct merge: 4
Problem sites with county recovered: 0


,_id_site,id_agency,_id_bank,FoodBank,_id_Agency,County,AgencyType,AgencySubType,City,Zip,isDistributionSite,block_geoid
0,15364.0,15364.0,3.0,Second Harvest Heartland,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,15366.0,15366.0,3.0,Second Harvest Heartland,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,15361.0,15361.0,3.0,Second Harvest Heartland,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,15365.0,15365.0,3.0,Second Harvest Heartland,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


HouseholdsReg: 48,315 unresolved out of 2,918,681 total (1.66%)
IndividualsReg: 137,064 unresolved out of 9,146,625 total (1.50%)
PoundsReg: 2,236,318 unresolved out of 157,814,745 total (1.42%)


In [28]:
#i stopped chasing the missing values, trying to understand what they actually are

unresolved_sites = q6_2024_clean[
    q6_2024_clean["County"].isna()
].copy()

cols = [
    "_id_site",
    "__ID",
    "_Year",
    "_Month",
    "FoodBank",
    "_id_bank",
    "id_agency",
    "AgencyType",
    "HouseholdsReg",
    "IndividualsReg",
    "PoundsReg",
    "zzCreationTS",
    "zzModificationTS"
]

display(
    unresolved_sites[cols]
    .sort_values(["_id_site", "_Month"])
)
summary_unresolved = (
    unresolved_sites
    .groupby("_id_site")
    .agg(
        months_reported=("_Month", "nunique"),
        foodbank=("FoodBank", lambda x: list(x.dropna().unique())),
        total_households=("HouseholdsReg", "sum"),
        total_individuals=("IndividualsReg", "sum"),
        total_pounds=("PoundsReg", "sum")
    )
    .reset_index()
)

display(summary_unresolved)

,_id_site,__ID,_Year,_Month,FoodBank,_id_bank,id_agency,AgencyType,HouseholdsReg,IndividualsReg,PoundsReg,zzCreationTS,zzModificationTS
12715,15361.0,25825,2024,1,Second Harvest Heartland,3.0,15361.0,NaN,620.0,1700.0,35948.00,2024-02-15T14:25:26Z,2025-12-30T13:50:32Z
13111,15361.0,26221,2024,2,Second Harvest Heartland,3.0,15361.0,NaN,570.0,1451.0,33360.00,2024-03-15T09:29:33Z,2025-12-30T13:50:32Z
13534,15361.0,26644,2024,3,Second Harvest Heartland,3.0,15361.0,NaN,472.0,1119.0,23416.00,2024-04-15T10:50:37Z,2025-12-30T13:50:32Z
14090,15361.0,27200,2024,4,Second Harvest Heartland,3.0,15361.0,NaN,575.0,1452.0,21666.00,2024-05-15T15:24:49Z,2025-12-30T13:50:32Z
14564,15361.0,27674,2024,5,Second Harvest Heartland,3.0,15361.0,NaN,680.0,1695.0,32934.00,2024-06-17T14:35:46Z,2025-12-30T13:50:33Z
...,...,...,...,...,...,...,...,...,...,...,...,...,...
13425,15430.0,26535,2024,2,NaN,NaN,NaN,NaN,82.0,173.0,2165.06,2024-03-16T12:23:51Z,2025-12-30T13:50:32Z
13416,15431.0,26526,2024,2,NaN,NaN,NaN,NaN,34.0,71.0,1590.00,2024-03-16T12:23:51Z,2025-12-30T13:50:32Z
13431,15432.0,26541,2024,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2024-03-16T12:23:51Z,2025-12-30T13:50:32Z
13406,15433.0,26516,2024,2,NaN,NaN,NaN,NaN,311.0,990.0,0.00,2024-03-16T12:23:51Z,2025-12-30T13:50:32Z


,_id_site,months_reported,foodbank,total_households,total_individuals,total_pounds
0,15361.0,7,[Second Harvest Heartland],3934.0,9945.0,205028.00
1,15364.0,7,[Second Harvest Heartland],1212.0,3325.0,44133.00
2,15365.0,7,[Second Harvest Heartland],6361.0,17079.0,393952.00
3,15366.0,7,[Second Harvest Heartland],30900.0,92540.0,1275578.00
4,15383.0,2,[],209.0,613.0,11157.00
5,15388.0,1,[],312.0,663.0,34692.00
6,15389.0,2,[],141.0,283.0,7000.00
7,15390.0,1,[],436.0,371.0,27251.00
8,15393.0,1,[],125.0,270.0,7051.00
9,15398.0,1,[],0.0,0.0,0.00


In [29]:
# Known-location 2024 records
known_2024 = q6_2024_clean[
    q6_2024_clean["County"].notna()
].copy()

# Missing-location 2024 records
missing_2024 = q6_2024_clean[
    q6_2024_clean["County"].isna()
].copy()

metrics = ["HouseholdsReg", "IndividualsReg", "PoundsReg"]

results = []

for _, row in missing_2024.iterrows():

    possible = known_2024[
        known_2024["_Month"] == row["_Month"]
    ].copy()

    for col in metrics:
        if pd.isna(row[col]):
            possible = possible[possible[col].isna()]
        else:
            possible = possible[possible[col] == row[col]]

    results.append({
        "_id_site": row["_id_site"],
        "_Month": row["_Month"],
        "HouseholdsReg": row["HouseholdsReg"],
        "IndividualsReg": row["IndividualsReg"],
        "PoundsReg": row["PoundsReg"],
        "exact_matches": len(possible),
        "matched_site": possible["_id_site"].tolist(),
        "matched_county": possible["County"].tolist(),
        "matched_type": possible["AgencyType"].tolist()
    })

duplicate_check = pd.DataFrame(results)

display(duplicate_check)

print(
    "Rows with at least one exact match:",
    (duplicate_check["exact_matches"] > 0).sum()
)

print(
    "Rows with no exact match:",
    (duplicate_check["exact_matches"] == 0).sum()
)

,_id_site,_Month,HouseholdsReg,IndividualsReg,PoundsReg,exact_matches,matched_site,matched_county,matched_type
0,15364.0,1,148.0,393.0,5556.0,1,[14719.0],[Rice],[Brick and Mortar Site]
1,15366.0,1,4188.0,12356.0,180317.0,1,[1279.0],[Steele],[Brick and Mortar Site]
2,15361.0,1,620.0,1700.0,35948.0,1,[14801.0],[Freeborn],[Brick and Mortar Site]
3,15365.0,1,913.0,2353.0,58330.0,1,[14072.0],[Rice],[Brick and Mortar Site]
4,15389.0,1,NaN,NaN,NaN,30,"[15286.0, 14694.0, 14620.0, 14372.0, 14392.0, ...","[Hubbard, Hennepin, Hennepin, Hennepin, Hennep...","[Meal Programs, Meal Programs, Meal Programs, ..."
...,...,...,...,...,...,...,...,...,...
59,15365.0,6,856.0,2360.0,54925.0,1,[14072.0],[Rice],[Brick and Mortar Site]
60,15366.0,7,5522.0,16876.0,218027.0,1,[1279.0],[Steele],[Brick and Mortar Site]
61,15364.0,7,131.0,311.0,5187.0,1,[14719.0],[Rice],[Brick and Mortar Site]
62,15361.0,7,470.0,1166.0,28852.0,1,[14801.0],[Freeborn],[Brick and Mortar Site]


Rows with at least one exact match: 58
Rows with no exact match: 6


In [30]:
# Only use rows where all three Q6 metrics are actually observed
dup_evidence = duplicate_check[
    duplicate_check[
        ["HouseholdsReg", "IndividualsReg", "PoundsReg"]
    ].notna().all(axis=1)
].copy()

print("Fully observed mystery rows:", len(dup_evidence))

print(
    "Fully observed rows with exactly one matching known site:",
    (dup_evidence["exact_matches"] == 1).sum()
)

print(
    "Fully observed rows with multiple matches:",
    (dup_evidence["exact_matches"] > 1).sum()
)

print(
    "Fully observed rows with no match:",
    (dup_evidence["exact_matches"] == 0).sum()
)
mapping_check = (
    dup_evidence[dup_evidence["exact_matches"] == 1]
    .assign(
        matched_site_one=lambda x: x["matched_site"].str[0],
        matched_county_one=lambda x: x["matched_county"].str[0]
    )
    .groupby("_id_site")
    .agg(
        months=("_Month", "count"),
        matched_sites=("matched_site_one", lambda x: list(pd.unique(x))),
        matched_counties=("matched_county_one", lambda x: list(pd.unique(x)))
    )
    .reset_index()
)

display(mapping_check)

Fully observed mystery rows: 60
Fully observed rows with exactly one matching known site: 51
Fully observed rows with multiple matches: 4
Fully observed rows with no match: 5


,_id_site,months,matched_sites,matched_counties
0,15361.0,7,[14801.0],[Freeborn]
1,15364.0,6,[14719.0],[Rice]
2,15365.0,7,[14072.0],[Rice]
3,15366.0,7,[1279.0],[Steele]
4,15383.0,1,[1118.0],[Itasca]
5,15388.0,1,[1248.0],[Crow Wing]
6,15389.0,1,[1117.0],[Crow Wing]
7,15393.0,1,[1131.0],[Crow Wing]
8,15400.0,1,[1137.0],[Koochiching]
9,15401.0,1,[1314.0],[Cass]


In [31]:
ambiguous = dup_evidence[
    dup_evidence["exact_matches"] != 1
][
    [
        "_id_site",
        "_Month",
        "HouseholdsReg",
        "IndividualsReg",
        "PoundsReg",
        "exact_matches",
        "matched_site",
        "matched_county",
        "matched_type"
    ]
]

display(ambiguous)

display(
    duplicate_check[
        duplicate_check["_id_site"] == 15364
    ]
)

,_id_site,_Month,HouseholdsReg,IndividualsReg,PoundsReg,exact_matches,matched_site,matched_county,matched_type
5,15364.0,2,256.0,996.0,8184.00,0,[],[],[]
12,15433.0,2,311.0,990.0,0.00,0,[],[],[]
14,15390.0,2,436.0,371.0,27251.00,0,[],[],[]
21,15399.0,2,966.0,2683.0,53145.00,0,[],[],[]
30,15404.0,2,261.0,710.0,14225.00,2,"[1189.0, 1189.0]","[Kanabec, Kanabec]","[Brick and Mortar Site, Brick and Mortar Site]"
31,15430.0,2,82.0,173.0,2165.06,2,"[14109.0, 1232.0]","[Itasca, Cass]","[Brick and Mortar Site, Brick and Mortar Site]"
34,15407.0,2,92.0,239.0,4593.00,2,"[1365.0, 1365.0]","[Kanabec, Kanabec]","[Brick and Mortar Site, Brick and Mortar Site]"
38,15421.0,2,82.0,173.0,2165.06,2,"[14109.0, 1232.0]","[Itasca, Cass]","[Brick and Mortar Site, Brick and Mortar Site]"
47,15383.0,3,86.0,247.0,3326.00,0,[],[],[]


,_id_site,_Month,HouseholdsReg,IndividualsReg,PoundsReg,exact_matches,matched_site,matched_county,matched_type
0,15364.0,1,148.0,393.0,5556.0,1,[14719.0],[Rice],[Brick and Mortar Site]
5,15364.0,2,256.0,996.0,8184.0,0,[],[],[]
44,15364.0,3,152.0,282.0,5693.0,1,[14719.0],[Rice],[Brick and Mortar Site]
48,15364.0,4,151.0,362.0,5709.0,1,[14719.0],[Rice],[Brick and Mortar Site]
52,15364.0,5,188.0,546.0,7469.0,1,[14719.0],[Rice],[Brick and Mortar Site]
56,15364.0,6,186.0,435.0,6335.0,1,[14719.0],[Rice],[Brick and Mortar Site]
61,15364.0,7,131.0,311.0,5187.0,1,[14719.0],[Rice],[Brick and Mortar Site]


In [32]:
known_unique = known_2024.drop_duplicates(
    subset=[
        "_id_site",
        "_Month",
        "HouseholdsReg",
        "IndividualsReg",
        "PoundsReg"
    ]
).copy()

# Compare 15364 with its likely counterpart 14719
display(
    q6_2024_clean[
        q6_2024_clean["_id_site"].isin([15364, 14719])
    ][
        [
            "_id_site",
            "_Month",
            "County",
            "HouseholdsReg",
            "IndividualsReg",
            "PoundsReg"
        ]
    ].sort_values(["_Month", "_id_site"])
)

display(
    q6_2024_clean[
        q6_2024_clean["_id_site"].isin([15383, 1118])
    ][
        [
            "_id_site",
            "_Month",
            "County",
            "HouseholdsReg",
            "IndividualsReg",
            "PoundsReg"
        ]
    ].sort_values(["_Month", "_id_site"])
)

,_id_site,_Month,County,HouseholdsReg,IndividualsReg,PoundsReg
12527,14719.0,1,Rice,148.0,393.0,5556.0
12711,15364.0,1,NaN,148.0,393.0,5556.0
12974,14719.0,2,Rice,150.0,428.0,5686.0
13063,15364.0,2,NaN,256.0,996.0,8184.0
13735,14719.0,3,Rice,152.0,282.0,5693.0
13531,15364.0,3,NaN,152.0,282.0,5693.0
14300,14719.0,4,Rice,151.0,362.0,5709.0
14084,15364.0,4,NaN,151.0,362.0,5709.0
14503,14719.0,5,Rice,188.0,546.0,7469.0
14542,15364.0,5,NaN,188.0,546.0,7469.0


,_id_site,_Month,County,HouseholdsReg,IndividualsReg,PoundsReg
17905,1118.0,1,Itasca,89.0,306.0,4806.0
17935,1118.0,2,Itasca,123.0,366.0,7831.0
13410,15383.0,2,NaN,123.0,366.0,7831.0
17975,1118.0,3,Itasca,86.0,247.0,2183.0
13762,15383.0,3,NaN,86.0,247.0,3326.0
18018,1118.0,4,Itasca,76.0,227.0,3205.0
18062,1118.0,5,Itasca,95.0,256.0,2998.0
18107,1118.0,6,Itasca,85.0,269.0,3307.0
18151,1118.0,7,Itasca,73.0,212.0,2745.0
18195,1118.0,8,Itasca,74.0,245.0,6316.0


In [33]:
# confirmed duplicate = same month + same households + same individuals + same pounds
confirmed_dups = duplicate_check[
    duplicate_check["exact_matches"] >= 1
].copy()

# Missing-location rows that have at least one exact known-location match
dup_ids = set(
    confirmed_dups.index
)

# Build a mask directly from duplicate_check row order
missing_2024_reset = missing_2024.reset_index(drop=True)

confirmed_mask = duplicate_check["exact_matches"] >= 1

unresolved_after_dedup = missing_2024_reset[
    ~confirmed_mask
].copy()

print("Remaining unresolved rows:", len(unresolved_after_dedup))

for col in ["HouseholdsReg", "IndividualsReg", "PoundsReg"]:
    unresolved_sum = unresolved_after_dedup[col].sum()
    total_sum = q6_2024_clean[col].sum()

    print(
        f"{col}: {unresolved_sum:,.0f} "
        f"({100 * unresolved_sum / total_sum:.3f}% of 2024 total)"
    )

Remaining unresolved rows: 6
HouseholdsReg: 2,111 (0.072% of 2024 total)
IndividualsReg: 5,388 (0.059% of 2024 total)
PoundsReg: 91,906 (0.058% of 2024 total)


In [36]:
def unique_sorted(values):
    values = values.dropna().unique()
    return sorted(values)

def unique_list(values):
    values = values.dropna().unique()
    return list(values)

site_type_history = (
    q6_clean[
        q6_clean["_Month"].between(1, 12)
    ]
    .groupby("_id_site")
    .agg(
        years=("_Year", unique_sorted),
        agency_types=("AgencyType", unique_list)
    )
    .reset_index()
)

display(site_type_history.head(20))
site_type_history["number_of_types"] = (
    site_type_history["agency_types"]
    .apply(len)
)

changed_type_sites = site_type_history[
    site_type_history["number_of_types"] > 1
]

display(changed_type_sites)

,_id_site,years,agency_types
0,1066.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]"
1,1067.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]"
2,1069.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]"
3,1070.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]"
4,1071.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]"
5,1072.0,"[2022, 2023, 2024, 2025, 2026]",[Brick and Mortar Site]
6,1073.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]"
7,1074.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]"
8,1075.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]"
9,1076.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]"


,_id_site,years,agency_types,number_of_types
0,1066.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]",2
1,1067.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]",2
2,1069.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]",2
3,1070.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]",2
4,1071.0,"[2022, 2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]",2
...,...,...,...,...
580,15372.0,"[2023, 2024, 2025, 2026]","[Food Shelf, Brick and Mortar Site]",2
581,15375.0,"[2023, 2024, 2025]","[Food Shelf, Outreach Program]",2
582,15376.0,"[2023, 2024, 2025]","[Food Shelf, Outreach Program]",2
583,15377.0,"[2023, 2024]","[Food Shelf, Outreach Program]",2


In [37]:
# Create indicator columns for broad site categories
site_rows = q6_clean[
    q6_clean["_Month"].between(1, 12)
].copy()

site_rows["is_core_shelf"] = site_rows["AgencyType"].isin([
    "Food Shelf",
    "Brick and Mortar Site",
    "Brick and Mortar Sites"
])

site_rows["is_mobile_outreach"] = site_rows["AgencyType"].isin([
    "Mobile",
    "Outreach Program"
])

site_rows["is_meal_program"] = site_rows["AgencyType"].isin([
    "Meal Program",
    "Meal Programs",
    "On-Site Meal Programs"
])

site_rows["is_college"] = (
    site_rows["AgencyType"] == "College Campus Program"
)

site_rows["is_tribal"] = (
    site_rows["AgencyType"] == "Tribal Program"
)

site_class = (
    site_rows
    .groupby("_id_site")[
        [
            "is_core_shelf",
            "is_mobile_outreach",
            "is_meal_program",
            "is_college",
            "is_tribal"
        ]
    ]
    .max()
    .reset_index()
)

display(site_class.head())

conditions = [
    site_class["is_core_shelf"] & site_class["is_mobile_outreach"],
    site_class["is_core_shelf"],
    site_class["is_mobile_outreach"],
    site_class["is_meal_program"],
    site_class["is_college"],
    site_class["is_tribal"]
]

choices = [
    "Food Shelf / Outreach",
    "Food Shelf",
    "Mobile / Outreach",
    "Meal Program",
    "College Program",
    "Tribal Program"
]

site_class["stable_type"] = np.select(
    conditions,
    choices,
    default="Other"
)

print(site_class["stable_type"].value_counts())

,_id_site,is_core_shelf,is_mobile_outreach,is_meal_program,is_college,is_tribal
0,1066.0,True,False,False,False,False
1,1067.0,True,False,False,False,False
2,1069.0,True,False,False,False,False
3,1070.0,True,False,False,False,False
4,1071.0,True,False,False,False,False


stable_type
Food Shelf               447
Meal Program              65
Other                     55
Mobile / Outreach         52
Food Shelf / Outreach     41
Tribal Program             4
College Program            1
Name: count, dtype: int64


In [38]:
other_ids = site_class.loc[
    site_class["stable_type"] == "Other",
    "_id_site"
]

other_rows = site_rows[
    site_rows["_id_site"].isin(other_ids)
]

other_types = (
    other_rows
    .groupby("_id_site")["AgencyType"]
    .unique()
    .reset_index()
)

display(other_types)

other_activity = (
    other_rows[
        ["HouseholdsReg", "IndividualsReg", "PoundsReg"]
    ]
    .sum()
)

print(other_activity)
print(
    other_rows["AgencyType"]
    .value_counts(dropna=False)
)

,_id_site,AgencyType
0,15288.0,[nan]
1,15361.0,[nan]
2,15362.0,[nan]
3,15363.0,[nan]
4,15364.0,[nan]
5,15365.0,[nan]
6,15366.0,[nan]
7,15383.0,[nan]
8,15388.0,[nan]
9,15389.0,[nan]


HouseholdsReg      127991.00
IndividualsReg     381374.00
PoundsReg         5891221.12
dtype: float64
AgencyType
NaN                        130
Agency or Tribal Nation     14
Name: count, dtype: int64


In [39]:
#removing the confirmed duplicate missing-location records
known_rows = q6_clean[
    q6_clean["County"].notna()
].copy()

missing_rows = q6_clean[
    q6_clean["County"].isna()
].copy()

# Keep only valid months
known_rows = known_rows[
    known_rows["_Month"].between(1, 12)
]

missing_rows = missing_rows[
    missing_rows["_Month"].between(1, 12)
]

# Remove exact duplicate known rows first
known_unique = known_rows.drop_duplicates(
    subset=[
        "_id_site",
        "_Year",
        "_Month",
        "HouseholdsReg",
        "IndividualsReg",
        "PoundsReg"
    ]
).copy()

In [ ]:
#inspect the 14 Agency or Tribal Nation sites
agency_tribal_ids = other_rows.loc[
    other_rows["AgencyType"] == "Agency or Tribal Nation",
    "_id_site"
].unique()

agency_tribal_info = org_data[
    org_data["__ID"].isin(agency_tribal_ids)
][
    [
        "__ID",
        "County",
        "AgencyType",
        "AgencySubType",
        "isDistributionSite",
        "isAgency",
        "isActive",
        "City",
        "Zip",
        "block_geoid"
    ]
]

display(agency_tribal_info)
print("Sites found:", agency_tribal_info["__ID"].nunique())

print("\nDistribution site values:")
print(
    agency_tribal_info["isDistributionSite"]
    .value_counts(dropna=False)
)

print("\nAgency subtypes:")
print(
    agency_tribal_info["AgencySubType"]
    .value_counts(dropna=False)
)

#based on result all agency or tribal are agencies/amin orgs
# i will exclude from food shelf activity

,__ID,County,AgencyType,AgencySubType,isDistributionSite,isAgency,isActive,City,Zip,block_geoid
910,16499.0,Hennepin,Agency or Tribal Nation,NaN,0.0,1.0,1.0,Minneapolis,55455.0,2.705310e+14
930,16681.0,Stearns,Agency or Tribal Nation,NaN,0.0,1.0,1.0,Albany,56307.0,2.714501e+14
940,16692.0,Hennepin-Suburbs,Agency or Tribal Nation,NaN,0.0,1.0,1.0,Brooklyn Center,55429.0,2.705302e+14
943,16695.0,Cass,Agency or Tribal Nation,NaN,0.0,1.0,1.0,Cass Lake,56633.0,2.702194e+14
951,16703.0,Crow Wing,Agency or Tribal Nation,NaN,0.0,1.0,1.0,Crosslake,56442.0,2.703595e+14
953,16705.0,Itasca,Agency or Tribal Nation,NaN,0.0,1.0,1.0,Deer River,56636.0,2.706148e+14
1001,16753.0,Cass,Agency or Tribal Nation,NaN,0.0,1.0,1.0,Pine River,56474.0,2.702196e+14
1006,16758.0,Cass,Agency or Tribal Nation,NaN,0.0,1.0,1.0,Remer,56672.0,NaN
1049,16801.0,Aitkin,Agency or Tribal Nation,NaN,0.0,1.0,1.0,Jacobson,55752.0,2.700177e+14
1050,16802.0,Aitkin,Agency or Tribal Nation,NaN,0.0,1.0,1.0,Aitkin,56431.0,2.700177e+14


Sites found: 14

Distribution site values:
isDistributionSite
0.0    14
Name: count, dtype: int64

Agency subtypes:
AgencySubType
NaN    14
Name: count, dtype: int64


In [41]:
# Keep the site-level fields we need from Organization.xlsx
org_site_info = org_data[
    [
        "__ID",
        "isDistributionSite",
        "isAgency",
        "isActive"
    ]
].copy()

# Merge Organization information onto our stable site classification
site_class_check = site_class.merge(
    org_site_info,
    left_on="_id_site",
    right_on="__ID",
    how="left"
)

# Compare our stable categories with the official distribution-site flag
distribution_table = pd.crosstab(
    site_class_check["stable_type"],
    site_class_check["isDistributionSite"],
    dropna=False
)

display(distribution_table)
missing_distribution_flag = (
    site_class_check[
        site_class_check["isDistributionSite"].isna()
    ]
    ["stable_type"]
    .value_counts()
)

print(missing_distribution_flag)

isDistributionSite,0.0,1.0,NaN
stable_type,,,
College Program,0,1,0
Food Shelf,0,446,1
Food Shelf / Outreach,0,41,0
Meal Program,0,65,0
Mobile / Outreach,0,52,0
Other,14,7,34
Tribal Program,0,4,0


stable_type
Other         34
Food Shelf     1
Name: count, dtype: int64


In [42]:
# Q6 activity variables
metrics = [
    "HouseholdsReg",
    "IndividualsReg",
    "PoundsReg"
]

# Keep only valid months for now
valid_month_data = q6_clean[
    q6_clean["_Month"].between(1, 12)
].copy()

# Known-location rows
known_all = valid_month_data[
    valid_month_data["County"].notna()
].copy()

# Missing-location rows
missing_all = valid_month_data[
    valid_month_data["County"].isna()
].copy()

# Only use rows where all 3 Q6 measures are reported
known_complete = known_all[
    known_all[metrics].notna().all(axis=1)
].copy()

missing_complete = missing_all[
    missing_all[metrics].notna().all(axis=1)
].copy()

# Give each missing row its original row number
missing_complete = (
    missing_complete
    .reset_index()
    .rename(columns={"index": "original_index"})
)

# Remove duplicate copies on the known side
known_unique = known_complete.drop_duplicates(
    subset=[
        "_id_site",
        "_Year",
        "_Month",
        "HouseholdsReg",
        "IndividualsReg",
        "PoundsReg"
    ]
).copy()

match_columns = [
    "_Year",
    "_Month",
    "HouseholdsReg",
    "IndividualsReg",
    "PoundsReg"
]

matches_all = missing_complete.merge(
    known_unique[
        [
            "_id_site",
            "County"
        ] + match_columns
    ],
    on=match_columns,
    how="left",
    suffixes=("_missing", "_known")
)

# Keep actual matches only
actual_matches = matches_all[
    matches_all["_id_site_known"].notna()
].copy()
match_counts = (
    actual_matches
    .groupby("original_index")["_id_site_known"]
    .nunique()
    .reset_index(name="exact_matches")
)

duplicate_check_all = missing_complete.merge(
    match_counts,
    on="original_index",
    how="left"
)

duplicate_check_all["exact_matches"] = (
    duplicate_check_all["exact_matches"]
    .fillna(0)
    .astype(int)
)

print("Complete missing-location rows:",
      len(duplicate_check_all))

print("\nExactly one match:",
      (duplicate_check_all["exact_matches"] == 1).sum())

print("Multiple matches:",
      (duplicate_check_all["exact_matches"] > 1).sum())

print("No exact match:",
      (duplicate_check_all["exact_matches"] == 0).sum())

year_summary = pd.crosstab(
    duplicate_check_all["_Year"],
    duplicate_check_all["exact_matches"]
)

display(year_summary)

Complete missing-location rows: 107

Exactly one match: 85
Multiple matches: 2
No exact match: 20


exact_matches,0,1,2
_Year,,,
2023,4,32,0
2024,5,53,2
2025,3,0,0
2026,8,0,0


In [43]:
# Invalid month records
invalid_month_rows = q6_clean[
    ~q6_clean["_Month"].between(1, 12)
].copy()

month0 = invalid_month_rows[
    invalid_month_rows["_Month"] == 0
].copy()

month15 = invalid_month_rows[
    invalid_month_rows["_Month"] == 15
].copy()

print("Month 0 rows:", len(month0))
print("Month 15 rows:", len(month15))

dec_2023 = q6_clean[
    (q6_clean["_Year"] == 2023) &
    (q6_clean["_Month"] == 12)
][
    [
        "_id_site",
        "HouseholdsReg",
        "IndividualsReg",
        "PoundsReg"
    ]
].copy()

month0_check = month0.merge(
    dec_2023,
    on="_id_site",
    how="left",
    suffixes=("_month0", "_dec2023")
)

for col in metrics:
    month0_check[col + "_same"] = (
        month0_check[col + "_month0"].fillna(-999999)
        ==
        month0_check[col + "_dec2023"].fillna(-999999)
    )

same_columns = [
    "HouseholdsReg_same",
    "IndividualsReg_same",
    "PoundsReg_same"
]

month0_check["all_three_same"] = (
    month0_check[same_columns].all(axis=1)
)

print("Month-0 rows matching December 2023 exactly:")
print(month0_check["all_three_same"].value_counts())

display(
    month0_check[
        [
            "_id_site",
            "HouseholdsReg_month0",
            "HouseholdsReg_dec2023",
            "IndividualsReg_month0",
            "IndividualsReg_dec2023",
            "PoundsReg_month0",
            "PoundsReg_dec2023",
            "all_three_same"
        ]
    ]
)

march_2024 = q6_clean[
    (q6_clean["_Year"] == 2024) &
    (q6_clean["_Month"] == 3)
][
    [
        "_id_site",
        "HouseholdsReg",
        "IndividualsReg",
        "PoundsReg"
    ]
].copy()

month15_check = month15.merge(
    march_2024,
    on="_id_site",
    how="left",
    suffixes=("_month15", "_mar2024")
)

display(month15_check)

Month 0 rows: 25
Month 15 rows: 1
Month-0 rows matching December 2023 exactly:
all_three_same
True    25
Name: count, dtype: int64


,_id_site,HouseholdsReg_month0,HouseholdsReg_dec2023,IndividualsReg_month0,IndividualsReg_dec2023,PoundsReg_month0,PoundsReg_dec2023,all_three_same
0,14810.0,54.0,54.0,129.0,129.0,1785.0,1785.0,True
1,1072.0,183.0,183.0,421.0,421.0,8778.0,8778.0,True
2,14717.0,332.0,332.0,755.0,755.0,11625.0,11625.0,True
3,1103.0,274.0,274.0,691.0,691.0,12128.0,12128.0,True
4,1088.0,691.0,691.0,1106.0,1106.0,40515.0,40515.0,True
5,1110.0,146.0,146.0,463.0,463.0,13535.0,13535.0,True
6,1108.0,97.0,97.0,202.0,202.0,5820.0,5820.0,True
7,1250.0,118.0,118.0,188.0,188.0,6948.0,6948.0,True
8,1410.0,653.0,653.0,1772.0,1772.0,37695.0,37695.0,True
9,1128.0,36.0,36.0,59.0,59.0,1295.0,1295.0,True


,__ID,_id_site,_Year,_Month,DateForSearch,zzCreationTS,zzModificationTS,AdultsReg,ChildrenReg,HouseholdsReg_month15,...,_id_bank,SuperShelf,_YearMonth,PoundsPerIndiv,IndividualsPerHH,id_agency,PoundsPerHH,HouseholdsReg_mar2024,IndividualsReg_mar2024,PoundsReg_mar2024
0,23361,1148.0,2023,15,2024-03-01,2023-09-18T11:26:15Z,2025-12-30T13:50:30Z,37.0,15.0,33.0,...,5.0,NaN,202315,30.0,2.0,16246.0,62.0,43.0,97.0,2902.4


In [45]:
site_1148 = q6_clean[
    q6_clean["_id_site"] == 1148
][
    [
        "__ID",
        "_Year",
        "_Month",
        "DateForSearch",
        "HouseholdsReg",
        "IndividualsReg",
        "PoundsReg"
    ]
].sort_values(["_Year", "_Month"])

display(site_1148)
same_activity = q6_clean[
    (q6_clean["_id_site"] == 1148) &
    (q6_clean["HouseholdsReg"] == 33) &
    (q6_clean["IndividualsReg"] == 69) &
    (q6_clean["PoundsReg"] == 2043)
]

display(same_activity)

,__ID,_Year,_Month,DateForSearch,HouseholdsReg,IndividualsReg,PoundsReg
582,13447,2022,1,2022-01-01,22.0,35.0,2020.00
464,13329,2022,2,2022-02-01,29.0,46.0,2483.00
1090,13955,2022,3,2022-03-01,23.0,34.0,2010.00
1662,14536,2022,4,2022-04-01,21.0,35.0,2351.00
1725,14781,2022,5,2022-05-01,20.0,28.0,1473.00
2453,15510,2022,6,2022-06-01,24.0,28.0,1552.00
3020,16080,2022,7,2022-07-01,25.0,41.0,1909.00
3369,16430,2022,8,2022-08-01,25.0,35.0,1720.00
4032,17139,2022,9,2022-09-01,16.0,29.0,1262.00
4721,17828,2022,10,2022-10-01,25.0,37.0,1895.00


,__ID,_id_site,_Year,_Month,DateForSearch,zzCreationTS,zzModificationTS,AdultsReg,ChildrenReg,HouseholdsReg,...,PoundsHoliday,PoundsTotal,AgencyType,_id_bank,SuperShelf,_YearMonth,PoundsPerIndiv,IndividualsPerHH,id_agency,PoundsPerHH
6889,19998,1148.0,2023,1,2023-01-01,2023-03-15T14:31:21Z,2025-12-30T13:50:28Z,37.0,15.0,33.0,...,NaN,2043.0,Food Shelf,5.0,NaN,202301,30.0,2.0,NaN,62.0
10251,23361,1148.0,2023,15,2024-03-01,2023-09-18T11:26:15Z,2025-12-30T13:50:30Z,37.0,15.0,33.0,...,NaN,2043.0,Brick and Mortar Site,5.0,NaN,202315,30.0,2.0,16246.0,62.0


In [46]:
# Remove the 26 invalid-month duplicate records
valid_month_mask = q6_clean["_Month"].between(1, 12)

q6_valid_months = q6_clean[
    valid_month_mask
].copy()

print("Original rows:", len(q6_clean))
print("Rows removed:", len(q6_clean) - len(q6_valid_months))
print("Rows remaining:", len(q6_valid_months))

print(
    "Invalid months remaining:",
    (~q6_valid_months["_Month"].between(1, 12)).sum()
)

Original rows: 28287
Rows removed: 26
Rows remaining: 28261
Invalid months remaining: 0


In [47]:
metrics = [
    "HouseholdsReg",
    "IndividualsReg",
    "PoundsReg"
]

# Rows with a known county and all 3 Q6 measures observed
known_complete = q6_valid_months[
    q6_valid_months["County"].notna() &
    q6_valid_months[metrics].notna().all(axis=1)
].copy()

# Rows with missing county and all 3 measures observed
missing_complete = q6_valid_months[
    q6_valid_months["County"].isna() &
    q6_valid_months[metrics].notna().all(axis=1)
].copy()

# Preserve original row index so we know which rows to remove
missing_complete = (
    missing_complete
    .reset_index()
    .rename(columns={"index": "source_index"})
)

match_columns = [
    "_Year",
    "_Month",
    "HouseholdsReg",
    "IndividualsReg",
    "PoundsReg"
]

exact_matches = missing_complete.merge(
    known_complete[
        ["_id_site", "County"] + match_columns
    ],
    on=match_columns,
    how="left",
    suffixes=("_missing", "_known")
)

confirmed_duplicates = exact_matches[
    exact_matches["_id_site_known"].notna()
].copy()

duplicate_indices = (
    confirmed_duplicates["source_index"]
    .drop_duplicates()
    .tolist()
)

print("Confirmed blank-county duplicate rows:",
      len(duplicate_indices))

q6_no_duplicates = q6_valid_months[
    ~q6_valid_months.index.isin(duplicate_indices)
].copy()

print("Rows before duplicate removal:",
      len(q6_valid_months))

print("Duplicate rows removed:",
      len(q6_valid_months) - len(q6_no_duplicates))

print("Rows remaining:",
      len(q6_no_duplicates))

remaining_missing = q6_no_duplicates[
    q6_no_duplicates["County"].isna()
].copy()

print("Remaining rows with missing county:",
      len(remaining_missing))

print("Unique unresolved sites:",
      remaining_missing["_id_site"].nunique())

print("\nBy year:")
print(
    remaining_missing["_Year"]
    .value_counts()
    .sort_index()
)

Confirmed blank-county duplicate rows: 87
Rows before duplicate removal: 28261
Duplicate rows removed: 87
Rows remaining: 28174
Remaining rows with missing county: 45
Unique unresolved sites: 16

By year:
_Year
2022     6
2023    18
2024     9
2025     3
2026     9
Name: count, dtype: int64


In [48]:
unresolved = q6_no_duplicates[
    q6_no_duplicates["County"].isna()
].copy()

print("Agency types:")
print(unresolved["AgencyType"].value_counts(dropna=False))

print("\nFood banks:")
print(unresolved["FoodBank"].value_counts(dropna=False))

for col in ["HouseholdsReg", "IndividualsReg", "PoundsReg"]:

    unresolved_total = unresolved[col].sum()
    overall_total = q6_no_duplicates[col].sum()

    percent = 100 * unresolved_total / overall_total

    print(
        col,
        ":",
        round(unresolved_total, 2),
        "(",
        round(percent, 3),
        "%)"
    )

unresolved_summary = (
    unresolved
    .groupby("_id_site")
    .agg(
        rows=("_Month", "count"),
        first_year=("_Year", "min"),
        last_year=("_Year", "max"),
        households=("HouseholdsReg", "sum"),
        individuals=("IndividualsReg", "sum"),
        pounds=("PoundsReg", "sum")
    )
    .reset_index()
)

display(unresolved_summary)

Agency types:
AgencyType
NaN                 22
Meal Program        19
Outreach Program     3
Food Shelf           1
Name: count, dtype: int64

Food banks:
FoodBank
Second Harvest Northland    19
Second Harvest Heartland    17
NaN                          8
North Central Food Bank      1
Name: count, dtype: int64
HouseholdsReg : 13662.0 ( 0.11 %)
IndividualsReg : 40124.0 ( 0.109 %)
PoundsReg : 512799.0 ( 0.078 %)


,_id_site,rows,first_year,last_year,households,individuals,pounds
0,15218.0,19,2022,2026,0.0,0.0,0.0
1,15362.0,1,2023,2023,1433.0,4048.0,72020.0
2,15363.0,1,2023,2023,1433.0,4048.0,7202.0
3,15364.0,2,2023,2024,420.0,1502.0,13974.0
4,15366.0,1,2023,2023,3752.0,8328.0,160744.0
5,15367.0,1,2023,2023,0.0,0.0,0.0
6,15383.0,1,2024,2024,86.0,247.0,3326.0
7,15389.0,2,2023,2024,1131.0,3103.0,0.0
8,15390.0,1,2024,2024,436.0,371.0,27251.0
9,15398.0,1,2024,2024,0.0,0.0,0.0


In [49]:
# Start from the data after:
# 1. removing invalid-month duplicates
# 2. removing confirmed blank-county duplicate feeds

site_month_clean = q6_no_duplicates.copy()

# Add the stable site classification we created earlier
site_month_clean = site_month_clean.merge(
    site_class[
        ["_id_site", "stable_type"]
    ],
    on="_id_site",
    how="left"
)

# Add official Organization flags
org_flags = org_data[
    [
        "__ID",
        "isDistributionSite",
        "isAgency",
        "isActive",
        "block_geoid"
    ]
].copy()

site_month_clean = site_month_clean.merge(
    org_flags,
    left_on="_id_site",
    right_on="__ID",
    how="left",
    suffixes=("", "_org")
)

# Flag whether geographic location is usable
site_month_clean["location_resolved"] = (
    site_month_clean["County"].notna()
)
print("Final cleaned rows:", len(site_month_clean))

print("\nYears:")
print(
    site_month_clean["_Year"]
    .value_counts()
    .sort_index()
)

print("\nLocation resolved:")
print(
    site_month_clean["location_resolved"]
    .value_counts()
)

print("\nStable site types:")
print(
    site_month_clean["stable_type"]
    .value_counts(dropna=False)
)

Final cleaned rows: 28174

Years:
_Year
2022    5959
2023    6511
2024    6173
2025    6023
2026    3508
Name: count, dtype: int64

Location resolved:
location_resolved
True     28129
False       45
Name: count, dtype: int64

Stable site types:
stable_type
Food Shelf               21574
Meal Program              2700
Food Shelf / Outreach     1966
Mobile / Outreach         1648
Tribal Program             219
Other                       57
College Program             10
Name: count, dtype: int64


In [50]:
from pathlib import Path

processed_folder = Path("../data/processed")
processed_folder.mkdir(parents=True, exist_ok=True)

site_month_clean.to_csv(
    processed_folder / "site_month_clean.csv",
    index=False
)
unresolved.to_csv(
    processed_folder / "unresolved_location_rows.csv",
    index=False
)

The original dataset contained 28,287 records. Twenty-six invalid-month records were confirmed as duplicate observations and removed. An additional 87 records with missing county information were confirmed as duplicate feeds of known-location records using exact agreement on year, month, households, individuals, and pounds. The final cleaned dataset contains 28,174 site-month records. Forty-five records across 16 sites still lack reliable county information; these represent approximately 0.11% of households, 0.11% of individuals, and 0.08% of pounds and are retained with location_resolved = False but excluded from county-level analyses.